### PR TITLE
Let's give a cmake try to build per tools

### PR DESCRIPTION
**Pull request by Sashan**
The change also removes usused variables I've found occasionally.

    To build performance tools on windows we must add missing pieces:
        - getopt()
        - basename()
        - strcasecmp()
    all thos implementations are homebrewed

    Let cmake to determine openssl version we use to build tools.

    In order to tell CMake which OpenSSL version to use we use
    env. variabe OPENSSL_ROOT_DIR which points to OpenSSL installation
    prefix with desired version.

    We build evp_fetch and providerdoall test tools for OpenSSL 3.*.
    We don't build those for 1.1.1.

    CMake seems to work fine on windows and OpenBSD.

    The OpenBSD binaries produced by build process use RPATH. There is no
    such thing on Windows. On windows you need to add your OPENSSL_ROOT_DIR\bin
    to your PATH. Otherwise the .exe file produced by build process won't work.

### CODE REVIEW

**Comment by nhorman** [comment]
the changes look reasonable, but you have a fixup as your first commit, not sure that will autosquash properly

**Comment by Sashan** [comment]
I've just force pushed repository with squashed changes.

bonus: it builds on MacOS too. Gave a try 1.1.1, 3.0 3.3 and 3.4-dev everything worked fine.

File: source/perflib/strcasecmp.h
Comment (levitte) [inline source/perflib/strcasecmp.h:14]: You might want to edit the name 😉 

Also, FYI, Windows does have [`_stricmp`](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/stricmp-wcsicmp-mbsicmp-stricmp-l-wcsicmp-l-mbsicmp-l?view=msvc-170).

File: source/CMakeLists.txt
Comment (levitte) [inline source/CMakeLists.txt:32]: I suggest making some requirements:

```suggestion
find_package(OpenSSL REQUIRED)
```

File: source/perflib/strcasecmp.h
Comment (levitte) [reply source/perflib/strcasecmp.h:14]: Er, *this is strcasecmp.h*, changing the guard macro name was perhaps not the right move 😉

**Comment by levitte** [comment]
I did a bit of hacking, and a much bigger instruction comment in source/CMakeLists.txt.  Mind of I simply push a fixup commit into this PR?

**Comment by levitte** [comment]
This got me to test the OpenSSLConfig.cmake we produce for using an OpenSSL build without installing it, and oh boy, I found a bug (fixed in #24918)

**Comment by Sashan** [comment]
> I did a bit of hacking, and a much bigger instruction comment in source/CMakeLists.txt. Mind of I simply push a fixup commit into this PR?

please, go ahead and push your changes in.

**Comment by levitte** [comment]
Done.  Don't forget to pull...

File: source/CMakeLists.txt
Comment (levitte) [inline source/CMakeLists.txt:110]: It would be more CMake way like this:

```suggestion
if( OPENSSL_VERSION VERSION_GREATER_EQUAL 3 )
```

File: source/CMakeLists.txt
Comment (levitte) [inline source/CMakeLists.txt:97]: This looks a bit misplaced before `find_package(OpenSSL REQUIRED)`.

On the other hand, not sure it's needed, 'cause the include directory information is passed through with `target_link_libraries({program} PRIVATE perf OpenSSL::Crypto)` anyway, as far as I've understood.

Comment (Sashan) [reply source/CMakeLists.txt:97]: It looks like it is needed. I did try those options on my system (MacOS):
   - removing it completely, breaks the build (because we can not include <openssl/crypto.h>
   - just moving it to new location (after `find_package()`) keeps things working
   - moving it to new location but changing it to `include_directories(PRIVATE . )` breaks the build (can not find openssl/crypto.h)

I will move it after call to `find_package()`

Comment (levitte) [reply source/CMakeLists.txt]: How about letting it remain in the location you had, just remove `${OPENSSL_INCLUDE_DIR}`?  I tried that on Linux just now, and have no issue at all.

File: source/CMakeLists.txt
Comment (levitte) [inline source/CMakeLists.txt:108]: I just noticed, PRIVATE isn't a `include_directories` keyword:

https://cmake.org/cmake/help/latest/command/include_directories.html

(when doing a verbose build, I noticed commands that included `-I/home/levitte/gitwrk/OpenSSL/perftools/source/PRIVATE`, that's how I found out)

**Comment by levitte** [comment]
BTW, if you want to see what's actually happening when building:

``` console
$ cmake --build ./build --verbose
```

File: source/CMakeLists.txt
Comment (Sashan) [reply source/CMakeLists.txt]: I think `include_directories()` should follow `find_package()`. If I put it back before call to `find_package()` I must run `cmake -S . -B build` twice on MacOS so it generates files properly. If I call `include_directories()` after calling to `find_package()`.  Then running `cmake -S . -B build` just one time works.

The `{OPENSSL_INCLUDE_DIR}` must stay there, otherwise build can not find openssl header files to include.

I could remove `PRIVATE` word.

**Comment by Sashan** [comment]
update CMakeList.txt so it can build evp_setpeer which has just landed to repository.
I'm also doing a minor tweak to evp_setpeer.c to fix build on windows.

File: source/perflib/basename.c
Comment (andrewkdinh) [inline source/perflib/basename.c:19]: nit: Why is this put on a new line?

**Review by andrewkdinh — APPROVED** [review]

File: source/perflib/basename.c
Comment (Sashan) [reply source/perflib/basename.c:19]: sorry about it, this is a *BSD/Solaris coding style which I'm used to. will fix it.

FYI: the advantage of this style is you can quickly find a function definition by matching `^function` regex.

**Comment by Sashan** [comment]
@levitte it would be nice to have those changes in so everyone can build tools on windows. thanks.

**Comment by levitte** [comment]
> @levitte it would be nice to have those changes in so everyone can build tools on windows. thanks.

I agree.  Might I encourage you to nudge others to review them?

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
ping @openssl/committers for second review

File: source/perflib/basename.c
Comment (levitte) [reply source/perflib/basename.c:10]:
``` cmake
if (WIN32)
    target_source(perf PUBLIC perflib/basename.c)
endif()
```

**Comment by levitte** [comment]
I pushed a fixup just now, changing CMakeLists.txt a little bit.

This moves all the dependencies on OpenSSL to the perf library, and also declares its public include directory.

With that, all executables only need to depend on the perf library, and the magic of CMake will take care of the rest.

**Comment by Sashan** [comment]
please hold your new approvals/merging the cmake does not work on windows yet.
this is the error I get when I try to build it
```
C:\Users\OpenSSL\work.openssl\perftools\source>cmake --build build-master-one
ninja: error: build.ninja:54: bad $-escape (literal $ must be written as $$)
  INCLUDES = -IC:\Users\OpenSSL\work.openssl\perftools\source\$(PROJECT_...
                                                              ^ near here
```

not sure if it is glitch in cmake or ninja on windows, this is triggered by addition of `target_sources()` I'm still investigating the issue. will continue on Sunday evening.

File: source/CMakeLists.txt
Comment (levitte) [inline source/CMakeLists.txt:121]: I can't believe I didn't notice this... 

```suggestion
target_include_directories(perf PUBLIC "${PROJECT_SOURCE_DIR}")
```

Comment (Sashan) [reply source/CMakeLists.txt]: Thank you, from error, I had no idea what's the cause. the error message did not ring a bell at all.

The changes in this PR are ready for review/merge. I could build tools on OpenBSD for master and 1.1.1.
The build did work on Windows and Mac-OS too.

**Review by levitte — APPROVED** [review]

File: source/perflib/getopt.c
Comment (t8m) [inline source/perflib/getopt.c:10]: Is this still needed?

File: source/perflib/strcasecmp.c
Comment (t8m) [inline source/perflib/strcasecmp.c:10]: As above

File: source/pkeyread.c
Comment (t8m) [inline source/pkeyread.c:19]: Could you please do the usual indentation here and in the other source files where you're adding this?

File: source/perflib/strcasecmp.c
Comment (t8m) [inline source/perflib/strcasecmp.c:16]: Actually this is not needed as `#define strcasecmp _stricmp` would work.

File: source/perflib/basename.c
Comment (t8m) [inline source/perflib/basename.c:18]: Nit: remove space after `*`

File: source/perflib/basename.c
Comment (t8m) [inline source/perflib/basename.c:22]: Both `\\` and `/` should be supported as directory separator.

**Review by t8m — CHANGES_REQUESTED** [review]

File: source/perflib/basename.c
Comment (tom-cosgrove-arm) [inline source/perflib/basename.c:30]: Do we need this blank line?

File: source/perflib/basename.c
Comment (tom-cosgrove-arm) [inline source/perflib/basename.c:35]: These two lines are the same as 43-44 - would it be better to just have this once, outside the `if-else`?

File: source/perflib/basename.c
Comment (tom-cosgrove-arm) [inline source/perflib/basename.c:41]: Or replace the last two lines with

```
rv = tmp + dirseplen;
```

File: source/CMakeLists.txt
Comment (tom-cosgrove-arm) [inline source/CMakeLists.txt:5]: typo: `sources`

File: source/CMakeLists.txt
Comment (tom-cosgrove-arm) [inline source/CMakeLists.txt:42]: `s/openSSL/OpenSSL/` (and on line 50)

File: source/perflib/basename.c
Comment (tom-cosgrove-arm) [inline source/perflib/basename.c:26]: (coding style) I don't think we normally use parentheses in `return` statements, do we?

File: source/perflib/basename.c
Comment (tom-cosgrove-arm) [inline source/perflib/basename.c:20]: (coding style) isn't it normally 4 spaces per indent, not TAB?

File: source/perflib/basename.c
Comment (tom-cosgrove-arm) [inline source/perflib/basename.c:15]: Is this a Windows-only variant, or is it a Windows-compatible variant?

**Review by tom-cosgrove-arm — COMMENTED** [review]
Looks good, some comments

File: source/perflib/basename.c
Comment (tom-cosgrove-arm) [reply source/perflib/basename.c]: The updated code supports directory separators of more than one character, but both `\\` and `/` are only single characters. Do we need support for > 1 character directory separators?

> windows only so we can rely on directory separator to be a backslash.

Windows APIs support `/` as a directory separator, so we should too.

However, supporting `\\` and `/`  (and only those two) might be better - again, you can pass something like `C:\Users\Tom/projects/openssl` to Windows APIs and it will work as you would expect

File: source/perflib/basename.c
Comment (Sashan) [reply source/perflib/basename.c:15]: this is implementation of basename(3) for windows. I could not find a basename() equivalent on windows. The basename is defined by POSIX.1 (I think).

File: source/x509storeissuer.c
Comment (t8m) [inline source/x509storeissuer.c:20]: Indentation needed here and in other files above where you added the same snippets.

File: source/perflib/basename.c
Comment (Sashan) [reply source/perflib/basename.c]: thanks for clarification. it makes sense now. I did not know about '/' on windows.

**Comment by Sashan** [comment]
The basename() is reworked so both separators ('/' and '\\') are recognized on windows. Also function behavior got closer to behavior found on [unix/linux](https://man7.org/linux/man-pages/man3/basename.3.html)

File: source/perflib/basename.c
Comment (t8m) [inline source/perflib/basename.c:57]: And here

**Review by t8m — CHANGES_REQUESTED** [review]

File: source/perflib/basename.h
Comment (Sashan) [reply source/perflib/basename.h:14]: I know this different to what's found everywhere. The reason I've decided to diverge is the return value in case path is NULL (or empty string). In this case the function does `return ".";` Therefore I prefer `const char *`.

Comment (t8m) [reply source/perflib/basename.h:14]: Then remove this special behavior? Do we need it for anything?

File: source/perflib/basename.c
Comment (tom-cosgrove-arm) [inline source/perflib/basename.c:26]: Please don't use `>` to compare pointers that might be `NULL` - while this often works, [it's technically undefined behaviour](https://stackoverflow.com/a/35252004/7761803)

File: source/perflib/basename.h
Comment (Sashan) [reply source/perflib/basename.h:14]: another option would be to use static buffer and return it as a result. this is the closest to libc implementations. It does not need to be threadsafe. perftools call it from main thread only.

Comment (t8m) [reply source/perflib/basename.h:14]: IMO returning NULL or the empty string that was passed in is fine.

Comment (tom-cosgrove-arm) [reply source/perflib/basename.h:14]: I concur with @t8m: since this implementation is for Windows only (as I understand it), any callers must be able to cope with the behaviour of the POSIX functions (since that's the one they may get), so it doesn't help to give this one different behaviour, however useful it may be.

**Comment by tom-cosgrove-arm** [comment]
@Sashan Err, looks like you committed `basename.o`

File: source/basename.o
Comment (levitte) [inline source/basename.o:1]: Drop this file, please

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Commits to be squashed when merging.

File: source/perflib/basename.h
Comment (tom-cosgrove-arm) [inline source/perflib/basename.h:15]: typo: impleementation. "posix" should probably also be `POSIX`

**Comment by Sashan** [comment]
> Commits to be squashed when merging.

I can squash commits and do a force push when I'll be doing s/posix/POSIX, would that be OK?

**Comment by tom-cosgrove-arm** [comment]
> I can squash commits and do a force push when I'll be doing s/posix/POSIX, would that be OK?

I don't know about others, but when reviewing, I personally find it easier if commits are only added, with the squashing done on merge

**Review by tom-cosgrove-arm — APPROVED** [review]

**Comment by tom-cosgrove-arm** [comment]
Approved and labels updated

**Comment by t8m** [comment]
Merged. Thank you.

**Pull request closed without merging**
